### PR TITLE
feat: allow offset mode pagination in entity list

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

This allows `queryEntities` to be done with `offset`. It enables:

* Possibility to share a link to a specific page in the catalog 
* Changing number of rows displayed in the catalog
* Jumping pages by selecting page number from dropdown

image

Rebased and enhanced version of #22958

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
@vinzscam for some eyes too

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
@vinzscam any chance to get this going forward?

**Comment by freben** [comment]
he's on parental, the rest of us will have to give it a peek

**Review by freben — COMMENTED** [review]
Alright one more question.

File: plugins/catalog-react/src/hooks/useEntityListProvider.tsx [hunk 1]
```diff
@@ -155,39 +164,70 @@ export const EntityListProvider = <EntityFilters extends DefaultEntityFilters>(
...
+      }
+    }
+
+    const offset =
+      typeof parsed.offset === 'string' && paginationMode === 'offset'
+        ? Number.parseInt(parsed.offset, 10)
+        : undefined;
 
     return {
       queryParameters: (parsed.filters ?? {}) as Record<
```
Comment (freben) [inline plugins/catalog-react/src/hooks/useEntityListProvider.tsx:202]: Do we want to exclude the limit and offset from this? Do their presence pollute something elsewhere? Maybe it's perfectly ok but the type does imply that it should only contain keys that match the installed filters

File: plugins/catalog-react/src/hooks/useEntityListProvider.tsx [hunk 2]
```diff
@@ -349,7 +397,15 @@ export const EntityListProvider = <EntityFilters extends DefaultEntityFilters>(
...
-  }, [enablePagination, outputState.pageInfo]);
+  }, [paginationMode, outputState.pageInfo]);
+
+  const setLimitCb = useCallback((newLimit: number) => {
+    setLimit(newLimit);
+  }, []);
+
+  const setOffsetCb = useCallback((newOffset: number) => {
+    setOffset(newOffset);
+  }, []);
```
Comment (freben) [inline plugins/catalog-react/src/hooks/useEntityListProvider.tsx:408]: `useState` already is guaranteed to return static setters; this pair of useCallbacks will have no effect

**Review by freben — COMMENTED** [review]
Alright getting there. Just small questions

File: plugins/catalog-react/api-report.md [hunk 3]
```diff
@@ -307,20 +307,34 @@ export type EntityListContextProps<
     prev?: () => void;
   };
   totalItems?: number;
+  limit: number;
+  offset?: number;
+  setLimit: (limit: number) => void;
+  setOffset?: (offset: number) => void;
```
[the remark this answers is not in the document]
Comment (drodil) [reply plugins/catalog-react/api-report.md:313]: Yes, it's not available in cursor mode

File: plugins/catalog-react/src/hooks/useEntityListProvider.tsx (hunk 1 above)
Comment (drodil) [reply plugins/catalog-react/src/hooks/useEntityListProvider.tsx:202]: I would say they allow users to share their specific view of the catalog with URL parameters so they are quite handy.

**Comment by drodil** [comment]
@freben ready for another round :)

File: .changeset/hip-poems-invent.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+'@backstage/catalog-client': patch
+'@backstage/plugin-catalog-backend': patch
+'@backstage/plugin-catalog-react': patch
+'@backstage/plugin-catalog': patch
```
Comment (freben) [inline .changeset/hip-poems-invent.md:5]:
```suggestion
'@backstage/catalog-client': minor
'@backstage/plugin-catalog-backend': minor
'@backstage/plugin-catalog-react': minor
'@backstage/plugin-catalog': minor
```

**Review by freben — APPROVED** [review]
Alright, let's just bump the changesets and let's go

**Pull request merged**